### PR TITLE
Handle type switching of LoadBalancer from Internal to Public and vice versa

### LINKED ISSUE #269: Inefficient handling of `Service` object's `LoadBalancer` type switching
### Description:
When a Service object's LoadBalancer type is switched from `internal` to `external` (or vice versa), the current implementation attempts to patch the LoadBalancer type. This approach can cause undesirable side effects if the underlying LoadBalancer implementation does not support such type switching.

### Steps to reproduce:
1. Create a Service object with a type set to `LoadBalancer` and `internal`.
2. Once the internal LoadBalancer is created, switch the Service object's type from `internal` to `external`.
3. Observe the LoadBalancer's behavior.

### Expected Behavior:
When switching the LoadBalancer type in a Service object, the old LoadBalancer instance should be deleted and a new one should be created.

### Actual Behavior:
The current approach patches the LoadBalancer type, which can lead to unintended consequences if the LoadBalancer's implementation does not support type switches.

### Proposed Solution:
1. Detect changes to the `LoadBalancer` type in the Service object.
2. If a change is detected, delete the existing LoadBalancer instance.
3. Create a new LoadBalancer instance with the new type.

### Benefits:
- Avoids potential incompatibilities and issues related to patching LoadBalancer types.
- Ensures more consistent behavior when switching between internal and external LoadBalancers.

### PR DESCRIPTION
**Pull request by kasabe28**
# Proposed Changes

- Update load balancer controller to wait until LoadBalancerStatus is updated with valid IP, when load balancer type is switched Fixes #269 
- Modify testcase for load balancer

### CODE REVIEW

File: pkg/cloudprovider/onmetal/load_balancer.go [hunk 1]
```diff
@@ -87,7 +87,28 @@ func (o *onmetalLoadBalancer) GetLoadBalancerName(ctx context.Context, clusterNa
...
+	}
+
 	loadBalancerName := getLoadBalancerNameForService(clusterName, service)
+
+	// get existing load balancer type
+	existingLoadBalancer := &networkingv1alpha1.LoadBalancer{}
+	var existingLoadBalancerType networkingv1alpha1.LoadBalancerType
+	if err := o.onmetalClient.Get(ctx, client.ObjectKey{Namespace: o.onmetalNamespace, Name: loadBalancerName}, existingLoadBalancer); err == nil {
+		existingLoadBalancerType = existingLoadBalancer.Spec.Type
+		if existingLoadBalancerType != desiredLoadBalancerType {
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer.go:105]: You can do the check for the existing type directly in the if statement. There is no need to assign this to a variable.

File: pkg/cloudprovider/onmetal/load_balancer.go [hunk 2]
```diff
@@ -87,7 +87,28 @@ func (o *onmetalLoadBalancer) GetLoadBalancerName(ctx context.Context, clusterNa
...
+	} else {
+		desiredLoadBalancerType = networkingv1alpha1.LoadBalancerTypePublic
+	}
+
 	loadBalancerName := getLoadBalancerNameForService(clusterName, service)
+
+	// get existing load balancer type
+	existingLoadBalancer := &networkingv1alpha1.LoadBalancer{}
+	var existingLoadBalancerType networkingv1alpha1.LoadBalancerType
+	if err := o.onmetalClient.Get(ctx, client.ObjectKey{Namespace: o.onmetalNamespace, Name: loadBalancerName}, existingLoadBalancer); err == nil {
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer.go:103]: You also need to handle the error case here: e.g. what happens when there is an error getting the LB resource. One error could be NotFound and the other one could be some communication/RBAC etc issue.

File: pkg/cloudprovider/onmetal/load_balancer.go [hunk 3]
```diff
@@ -156,37 +177,40 @@ func (o *onmetalLoadBalancer) EnsureLoadBalancer(ctx context.Context, clusterNam
...
+	return &lbStatus, nil
 }
 
 func getLoadBalancerNameForService(clusterName string, service *v1.Service) string {
 	nameSuffix := strings.Split(string(service.UID), "-")[0]
 	return fmt.Sprintf("%s-%s-%s", clusterName, service.Name, nameSuffix)
 }
 
-func waitLoadBalancerActive(ctx context.Context, clusterName string, service *v1.Service, onmetalClient client.Client, loadBalancer *networkingv1alpha1.LoadBalancer) error {
+func waitLoadBalancerActive(ctx context.Context, existingLoadBalancerType networkingv1alpha1.LoadBalancerType, service *v1.Service,
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer.go:192]: Can we move the client as the second parameter of the method.

File: pkg/cloudprovider/onmetal/load_balancer.go [hunk 4]
```diff
@@ -156,37 +177,40 @@ func (o *onmetalLoadBalancer) EnsureLoadBalancer(ctx context.Context, clusterNam
...
 	backoff := wait.Backoff{
 		Duration: waitLoadbalancerInitDelay,
 		Factor:   waitLoadbalancerFactor,
 		Steps:    waitLoadbalancerActiveSteps,
 	}
 
+	loadBalancerStatus := v1.LoadBalancerStatus{}
 	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
-		err := onmetalClient.Get(ctx, client.ObjectKey{Namespace: service.Namespace, Name: loadBalancer.Name}, loadBalancer)
+		err := onmetalClient.Get(ctx, client.ObjectKey{Namespace: loadBalancer.Namespace, Name: loadBalancer.Name}, loadBalancer)
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer.go:203]: Can we use a single line `if err := ...` here.

File: pkg/cloudprovider/onmetal/load_balancer.go [hunk 5]
```diff
@@ -195,11 +219,11 @@ func waitLoadBalancerActive(ctx context.Context, clusterName string, service *v1
 	})
 
 	if wait.Interrupted(err) {
-		err = fmt.Errorf("timeout waiting for the onmetal LoadBalancer %s to become ready", client.ObjectKeyFromObject(loadBalancer))
+		return loadBalancerStatus, fmt.Errorf("timeout waiting for the onmetal LoadBalancer %s to become ready", client.ObjectKeyFromObject(loadBalancer))
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer.go:222]:
```suggestion
		return loadBalancerStatus, fmt.Errorf("timeout waiting for the LoadBalancer %s to become ready", client.ObjectKeyFromObject(loadBalancer))
```

File: pkg/cloudprovider/onmetal/load_balancer.go [hunk 6]
```diff
@@ -332,6 +356,33 @@ func (o *onmetalLoadBalancer) EnsureLoadBalancerDeleted(ctx context.Context, clu
...
 	}
+	if err := waitForDeletingLoadBalancer(ctx, service, o.onmetalClient, loadBalancer); err != nil {
+		return err
+	}
 	klog.V(2).InfoS("Deleted LoadBalancer", "LoadBalancer", client.ObjectKeyFromObject(loadBalancer))
 	return nil
 }
+
+func waitForDeletingLoadBalancer(ctx context.Context, service *v1.Service, onmetalClient client.Client, loadBalancer *networkingv1alpha1.LoadBalancer) error {
+	klog.V(2).InfoS("Waiting for onmetal LoadBalancer to be deleted", "LoadBalancer", client.ObjectKeyFromObject(loadBalancer))
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer.go:367]:
```suggestion
	klog.V(2).InfoS("Waiting for LoadBalancer instance to be deleted", "LoadBalancer", client.ObjectKeyFromObject(loadBalancer))
```

File: pkg/cloudprovider/onmetal/load_balancer.go [hunk 7]
```diff
@@ -332,6 +356,33 @@ func (o *onmetalLoadBalancer) EnsureLoadBalancerDeleted(ctx context.Context, clu
...
+			return true, nil
+		}
+		return false, err
+	})
+
+	if wait.Interrupted(err) {
+		return fmt.Errorf("timeout waiting for the onmetal LoadBalancer %s to be deleted", client.ObjectKeyFromObject(loadBalancer))
+	}
+
+	klog.V(2).InfoS("Deleted onmetal LoadBalancer", "LoadBalancer", client.ObjectKeyFromObject(loadBalancer))
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer.go:386]:
```suggestion
	klog.V(2).InfoS("Deleted LoadBalancer", "LoadBalancer", client.ObjectKeyFromObject(loadBalancer))
```

File: pkg/cloudprovider/onmetal/load_balancer_test.go [hunk 8]
```diff
@@ -272,45 +267,41 @@ var _ = Describe("LoadBalancer", func() {
...
 		loadBalancer := &networkingv1alpha1.LoadBalancer{
 			ObjectMeta: metav1.ObjectMeta{
 				Namespace: ns.Name,
 				Name:      lbProvider.GetLoadBalancerName(ctx, clusterName, service),
 			},
 		}
-		Eventually(Object(loadBalancer)).Should(SatisfyAll(
-			HaveField("Spec.Type", Equal(networkingv1alpha1.LoadBalancerTypeInternal))))
-
-		By("patching internal IP in load balancer status")
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer_test.go:285]: Is there a reason why the `By` statement has been removed?

File: pkg/cloudprovider/onmetal/load_balancer_test.go [hunk 9]
```diff
@@ -449,31 +443,26 @@ var _ = Describe("LoadBalancer", func() {
...
 		loadBalancer := &networkingv1alpha1.LoadBalancer{
 			ObjectMeta: metav1.ObjectMeta{
 				Namespace: ns.Name,
 				Name:      lbProvider.GetLoadBalancerName(ctx, clusterName, service),
 			},
 		}
-		Eventually(Object(loadBalancer)).Should(SatisfyAll(
-			HaveField("Spec.Type", Equal(networkingv1alpha1.LoadBalancerTypePublic))))
-
-		By("patching internal IP in load balancer status")
```
Comment (afritzler) [inline pkg/cloudprovider/onmetal/load_balancer_test.go:467]: Same as above.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/cloudprovider/onmetal/load_balancer.go (hunk 1 above)
Comment (kasabe28) [reply pkg/cloudprovider/onmetal/load_balancer.go:105]: `existingLoadBalancerType` is being passed to `waitLoadBalancerActive`, so variable is needed here.

File: pkg/cloudprovider/onmetal/load_balancer_test.go (hunk 8 above)
Comment (kasabe28) [reply pkg/cloudprovider/onmetal/load_balancer_test.go:285]: It is moved to line 270

File: pkg/cloudprovider/onmetal/load_balancer_test.go (hunk 9 above)
Comment (kasabe28) [reply pkg/cloudprovider/onmetal/load_balancer_test.go:467]: Moved to line 451

**Review by afritzler — APPROVED** [review]

**Pull request merged**